sycl/plugins/unified_runtime/ur/usm_allocator.cpp

```diff
@@ -855,27 +585,34 @@ bool Bucket::CanPool(bool &ToPool) {
   else
     NewFreeSlabsInBucket = AvailableSlabs.size() + 1;
   if (Capacity() >= NewFreeSlabsInBucket) {
-    size_t NewPoolSize = USMSettings.CurPoolSize + SlabAllocSize();
-    if (USMSettings.MaxPoolSize >= NewPoolSize) {
-      USMSettings.CurPoolSize = NewPoolSize;
-      if (chunkedBucket)
-        ++chunkedSlabsInPool;
-
-      updateStats(-1, 1);
-      ToPool = true;
-      return true;
+    size_t PoolSize = OwnAllocCtx.getParams().limits->TotalSize;
+    while (true) {
+      size_t NewPoolSize = PoolSize + SlabAllocSize();
+
+      if (OwnAllocCtx.getParams().limits->MaxSize < NewPoolSize) {
+        break;
+      }
+
+      if (OwnAllocCtx.getParams().limits->TotalSize.compare_exchange_strong(
+              PoolSize, NewPoolSize)) {
+        if (chunkedBucket)
+          ++chunkedSlabsInPool;
```

Comment:
how is this protected now that the lock at the beginning of this function was removed?

Reply:
My assumption was that the `PoolLock` was here only to protect `CurPoolSize` update (since this is a global variable, shared between all instances of USMAllocContext. This would match the comment: https://github.com/intel/llvm/blob/sycl/sycl/plugins/unified_runtime/ur/usm_allocator.cpp#L62 Now, instead of this lock I just rely on CAS to update the value.

As for the remaining variables (e.g. chunkedSlabsInPool) they should be protected by BucketLock which must be taken before calling `CanPool`

Reply:
thanks @igchor. if that's the case, then yes, we would be ok.